src/codec.cairo

```diff
@@ -25,70 +27,716 @@ pub impl EncodeSpanImpl<T, +Encode<@T>> of Encode<Span<T>> {
     }
 }
 
-pub impl EncodeByteArrayImpl of Encode<@ByteArray> {
+pub impl EncodeSpanImpl<T, +EncodeTrait<T>, +Copy<T>> of EncodeTrait<Span<T>> {
+    fn encode_to(self: Span<T>, ref dest: ByteArray) {
+        encode_compact_size(self.len(), ref dest);
+        for item in self {
+            (*item).encode_to(ref dest);
+        }
+    }
+}
+
+pub impl EncodeByteArrayImpl of EncodeTrait<@ByteArray> {
     fn encode_to(self: @ByteArray, ref dest: ByteArray) {
         encode_compact_size(self.len(), ref dest);
         dest.append(self);
     }
 }
 
+pub impl EncodeU32 of EncodeTrait<u32> {
+    fn encode_to(self: u32, ref dest: ByteArray) {
+        dest.append_word_rev(self.into(), 4);
+    }
+}
+
+pub impl EncodeU64 of EncodeTrait<u64> {
+    fn encode_to(self: u64, ref dest: ByteArray) {
+        dest.append_word_rev(self.into(), 8);
+    }
+}
+
+pub impl EncodeHash of EncodeTrait<Hash> {
+    fn encode_to(self: Hash, ref dest: ByteArray) {
+        dest.append(@self.into());
+    }
+}
+
+pub impl EncodeTxIn of EncodeTrait<TxIn> {
+    fn encode_to(self: TxIn, ref dest: ByteArray) {
+        self.previous_output.encode_to(ref dest);
+        self.script.encode_to(ref dest);
+        self.sequence.encode_to(ref dest);
+    }
+}
+
+pub impl EncodeTxOut of EncodeTrait<TxOut> {
+    fn encode_to(self: TxOut, ref dest: ByteArray) {
+        self.value.encode_to(ref dest);
+        self.pk_script.encode_to(ref dest);
+    }
+}
+
+pub impl EncodeOutpoint of EncodeTrait<OutPoint> {
+    fn encode_to(self: OutPoint, ref dest: ByteArray) {
+        self.txid.encode_to(ref dest);
+        self.vout.encode_to(ref dest);
+    }
+}
+
+pub impl EncodeTx of EncodeTrait<Transaction> {
+    fn encode_to(self: Transaction, ref dest: ByteArray) {
+        (self.version).encode_to(ref dest);
```

Comment:
Ig brackets are not required?

Anyways, once you have is_segwit flag here, you will need to conditionally append marker/flag and witnesses